lib/ts/nextjs/middleware.ts

```diff
@@ -0,0 +1,204 @@
+        } else if (tokens.antiCsrfToken.header) {
+            finalResponse.headers.append(ANTI_CSRF_TOKEN_HEADER_NAME, tokens.antiCsrfToken.header);
+        }
+        finalResponse.headers.append(
+            "set-cookie",
+            `${REDIRECT_ATTEMPT_COUNT_COOKIE_NAME}=${
+                redirectAttemptNumber + 1
+            }; Path=/; Max-Age=10; HttpOnly; SameSite=Strict`
+        );
+        logDebugMessage("Attached new tokens to response");
+        return finalResponse;
+    } catch (err) {
+        logDebugMessage("Error refreshing session");
+        logDebugMessage(err as unknown as string);
+        return redirectToAuthPage(request);
+    }
+}
+
+export async function revokeSession(config: SuperTokensNextjsConfig): Promise<Response | void> {
+    AppInfo = config.appInfo;
+    if (config.enableDebugLogs) {
+        enableLogging();
+    }
+    const signOutURL = new URL(`${AppInfo.apiBasePath}/signout`, AppInfo.apiDomain);
+    try {
+        const signOutResponse = await fetch(signOutURL, {
+            method: "POST",
+            headers: {
+                "Content-Type": "application/json",
+            },
+            credentials: "include",
+        });
+
+        if (signOutResponse.status !== 200) {
+            return;
+        }
+
+        const response = new Response(null, {});
+        response.headers.set("x-middleware-next", "1");
+        // TODO: Delete only the cookies that are auth related
+        response.headers.delete("set-cookie");
+        response.headers.delete(ACCESS_TOKEN_HEADER_NAME);
+        response.headers.delete(REFRESH_TOKEN_HEADER_NAME);
+        response.headers.delete(FRONT_TOKEN_HEADER_NAME);
+        response.headers.delete(ANTI_CSRF_TOKEN_HEADER_NAME);
+        return response;
+    } catch (err) {
+        logDebugMessage("Error revoking session");
+        logDebugMessage(err as unknown as string);
+        return;
+    }
+}
+
+function redirectToAuthPage(request: Request): Response {
+    const authPagePath = AppInfo.websiteBasePath || "/auth";
+    const redirectUrl = new URL(authPagePath, request.url);
+    redirectUrl.searchParams.set(FORCE_LOGOUT_PATH_PARAM_NAME, "true");
+    logDebugMessage(`Redirecting to: ${redirectUrl}`);
+    return new Response(null, {
+        status: 307,
+        headers: {
+            Location: redirectUrl.toString(),
+        },
+    });
+}
+
+async function fetchNewTokens(currentRefreshToken: string): Promise<{
+    accessToken: SuperTokensRequestToken;
+    refreshToken: SuperTokensRequestToken;
+    frontToken: SuperTokensRequestToken;
+    antiCsrfToken: SuperTokensRequestToken;
+}> {
+    const refreshApiURL = new URL(`${AppInfo.apiBasePath}/session/refresh`, AppInfo.apiDomain);
+    const refreshResponse = await fetch(refreshApiURL, {
+        method: "POST",
+        headers: {
+            "Content-Type": "application/json",
+            Cookie: `sRefreshToken=${currentRefreshToken}`,
```

Comment:
this isn't always allowed, we should set it in a cookie only if it came from a cookie.

Reply:
resolved